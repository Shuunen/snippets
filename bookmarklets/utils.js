/* eslint-disable-next-line no-unused-vars */
class Shuutils {
  constructor (app) {
    app = app || {
      id: 'shu-app',
      debug: true
    }
    this.app = app
    this.accentsIn = 'ÀÁÂÃÄÅĄàáâãäåąßÒÓÔÕÕÖØÓòóôõöøóÈÉÊËĘèéêëęðÇĆçćÐÌÍÎÏìíîïÙÚÛÜùúûüÑŃñńŠŚšśŸÿýŽŻŹžżź'
    this.accentsOut = 'AAAAAAAaaaaaaaBOOOOOOOOoooooooEEEEEeeeeeeCCccDIIIIiiiiUUUUuuuuNNnnSSssYyyZZZzzz'
  }

  log (...stuff) {
    if (!this.app.debug) {
      return
    }
    stuff.unshift(this.app.id + ' :')
    console.log.apply(console, stuff)
  }

  warn (...stuff) {
    stuff.unshift(this.app.id + ' :')
    console.warn.apply(console, stuff)
  }

  error (...stuff) {
    stuff.unshift(this.app.id + ' :')
    console.error.apply(console, stuff)
  }

  readableString (str) {
    return str.split('').map(function (letter) {
      var i = this.accentsIn.indexOf(letter)
      return i !== -1 ? this.accentsOut[i] : letter
    }).join('').replace(/(\W|-)/gi, ' ').replace(/\s+/, ' ')
  }

  debounce (func, wait, immediate) {
    var timeout
    return function () {
      var context = this
      var args = arguments
      var later = function later () {
        timeout = null
        if (!immediate) {
          func.apply(context, args)
        }
      }
      var callNow = immediate && !timeout
      clearTimeout(timeout)
      timeout = setTimeout(later, wait)
      if (callNow) {
        func.apply(context, args)
      }
    }
  }

  findOne (selector, context, dontYell) {
    context = context || document
    var item = context.querySelector(selector)
    if (item && this.app.debug) {
      this.log('found element matching "' + selector + '"')
    } else if (!item && !dontYell) {
      this.warn('found no element for selector "' + selector + '"')
    }
    return item
  }

  findFirst (selector, context, dontYell) {
    return this.findAll(selector, context, dontYell)[0]
  }

  findAll (selector, context, dontYell) {
    if (!selector || !selector.length || selector.length === 1) {
      this.error('incorrect selector : ', selector)
    }
    context = context || document
    var items = Array.prototype.slice.call(context.querySelectorAll(selector))
    if (items.length && this.app.debug) {
      this.log('found', items.length, 'elements matching "' + selector + '"')
    } else if (!items.length && !dontYell) {
      this.warn('found no elements for selector "' + selector + '"')
    }
    return items
  }
}
