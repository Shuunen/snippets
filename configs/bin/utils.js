import { copyFile, mkdir } from 'node:fs/promises'
import path from 'node:path'

/**
 * Get the filename from a filepath
 * @param {string} filepath the filepath to get the filename from
 * @returns {string} the filename
 * @example filename('C:\\Users\\me\\file.txt') // 'file.txt'
 */
export function filename(filepath = '') {
  return /[/\\](?<name>[\w.-]+)$/u.exec(filepath)?.groups?.name ?? ''
}

/**
 * Remove lines matching a regexp list
 * @param {string} content the content to clean
 * @param {RegExp[]} regexList the regex list to match
 * @returns {string} the cleaned content
 */
export function removeLinesMatching(content, regexList) {
  const lines = content.split('\n')
  const filteredLines = lines.filter(line => !regexList.some(regex => regex.test(line)))
  return filteredLines.join('\n').trim()
}

/**
 * Remove lines after a given regex
 * @param {string} content the content to remove lines from
 * @param {RegExp} regex the regexp to match the line after which to remove content
 * @returns {string} the content without the lines after the matching line
 */
export function removeLinesAfter(content, regex) {
  const lines = content.split('\n')
  const index = lines.findIndex(line => regex.test(line))
  if (index === -1) return content.trim()
  return lines.slice(0, index).join('\n').trim()
}

/**
 * Convert carriage return to unix line endings
 * @param {string} content the content to be processed
 * @returns the processed content with unix line endings
 */
export function useUnixCarriageReturn(content) {
  return content.replace(/\r\n/gu, '\n')
}

/**
 * Clean a file details content
 * @param {string} content the file content to clean
 * @param {RegExp} [linesAfter] a regex to remove lines after
 * @param {RegExp[]} [linesMatching] a list of regex to remove lines matching
 * @param shouldClearSpaces if true will also clear spaces
 * @returns {string} the cleaned file content
 */
// eslint-disable-next-line @typescript-eslint/max-params
export function clean(content, linesAfter, linesMatching, shouldClearSpaces = true) {
  if (!content) return ''
  let output = content
  if (linesAfter) output = removeLinesAfter(output, linesAfter)
  if (linesMatching) output = removeLinesMatching(output, linesMatching)
  if (shouldClearSpaces) output = output.replace(/\s*/gu, '')
  return output
}

/* c8 ignore start */

/**
 * Normalize a filepath with slash style
 * @param {string} filepath the filepath
 * @param {boolean} shouldUseTilde use tilde will replace the home directory with ~
 * @param {string} home the home directory path
 * @returns the normalized path
 */
export function normalizePathWithSlash(filepath, shouldUseTilde = false, home = process.env.HOME ?? '') {
  let outPath = path.normalize(filepath).replace(/\\/gu, '/')
  if (shouldUseTilde) outPath = outPath.replace(normalizePathWithSlash(home), '~')
  return outPath
}

/**
 * Copy a file
 * @param {string} source the source file
 * @param {string} destination the destination file
 * @returns {Promise<boolean>} some bool result; i dont know im in the train to Paris
 */
export async function copy(source, destination) {
  // destination will be created or overwritten by default.
  const destinationFolder = destination.replace(filename(destination), '')
  // eslint-disable-next-line @typescript-eslint/naming-convention
  await mkdir(destinationFolder, { recursive: true })

  return copyFile(source, destination)
    .then(() => true)
    .catch(error => {
      console.log(error)
      return false
    })
}
/* c8 ignore stop */
