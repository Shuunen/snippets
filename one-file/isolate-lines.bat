@echo off

node isolate-lines.js --clipboard
