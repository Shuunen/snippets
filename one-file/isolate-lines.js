#!/usr/bin/env node
const mergeUnique = (array) => [...new Set(array)]

export const isolateLines = (list) => {
  const lines = mergeUnique(list.replace(/ /g, '').split('\n')).filter(line => line.length > 0)
  return lines.sort().join('\n\n').trim()
}

const processClipboard = async () => {
  const clipboard = await import('clipboardy')
  const input = clipboard.default.readSync()
  const output = isolateLines(input)
  /* c8 ignore next */
  console.log('will copy this to clipboard :\n---\n' + output + `\n--- ${Date.now()}`)
  clipboard.default.writeSync(output)
}

/* c8 ignore next */
if (process.argv.includes('--clipboard')) processClipboard()
