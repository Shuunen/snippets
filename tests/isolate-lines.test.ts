import { expect, it } from 'vitest'
import { isolateLines } from '../one-file/isolate-lines.utils'

it('isolate lines A', () => {
  expect(isolateLines('xyz\nudp://9.7zip.t0:2750  xyz  \n \n \n http://ubuntu.com:80/announce  \n  ab-cd\n\n  ab-cd ')).toMatchSnapshot()
})

it('isolate lines B', () => { expect(isolateLines('')).toEqual([]) })

it('isolate lines C', () => { expect(isolateLines('  ')).toEqual([]) })

it('isolate lines D', () => { expect(isolateLines('  \n  ')).toEqual([]) })

it('isolate lines E', () => { expect(isolateLines('  \n  \n  ')).toEqual([]) })

it('isolate lines F', () => { expect(isolateLines('  \nxyz  \n  \n  ')).toEqual(['xyz']) })

