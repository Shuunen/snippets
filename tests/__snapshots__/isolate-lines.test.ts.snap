// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`isolate lines A 1`] = `
[
  "ab-cd",
  "http://ubuntu.com:80/announce",
  "udp://9.7zip.t0:2750xyz",
  "xyz",
]
`;
